### ISSUE TITLE
关于训练收敛的问题

### ISSUE THREAD

**Issue by stillwaterman** [thuml/Nonstationary_Transformers#6]
您好，我在ETTh2上测试您的ns_transformer模型，训练中我发现模型好像大部分时候在完成第一个epoch时val loss就最低了，但是train loss不是最低，于是之后的训练，train loss下降，val loss在最低点上方震荡，3个epoch后early stopping了，请问这是正常现象吗，模型会不会存在欠拟合的情况。下面附上我的实验情况。
@WenWeiTHU
Args in experiment:
Namespace(activation='gelu', batch_size=32, c_out=7, checkpoints='./checkpoints/', d_ff=2048, d_layers=1, d_model=512, data='ETTh2', data_path='ETTh2.csv', dec_in=7, des="'Exp_h256_l2'", devices='0,1,2,3', distil=True, do_predict=False, dropout=0.05, e_layers=2, embed='timeF', enc_in=7, factor=1, features='M', freq='h', gpu=0, is_training=1, itr=3, label_len=48, learning_rate=0.0001, loss='mse', lradj='type1', model='ns_Transformer', model_id='ETTh2_96_96', moving_avg=25, n_heads=8, num_workers=0, output_attention=False, p_hidden_dims=[256, 256], p_hidden_layers=2, patience=3, pred_len=96, root_path='./dataset/ETT-small/', seed=2021, seq_len=96, target='OT', train_epochs=10, use_amp=False, use_gpu=True, use_multi_gpu=False)
Use GPU: cuda:0
>>>>>>>start training : ETTh2_96_96_ns_Transformer_ETTh2_ftM_sl96_ll48_pl96_dm512_nh8_el2_dl1_df2048_fc1_ebtimeF_dtTrue_'Exp_h256_l2'_0>>>>>>>>>>>>>>>>>>>>>>>>>>
train 8449
val 2785
test 2785
	iters: 100, epoch: 1 | loss: 0.3959244
	speed: 0.0768s/iter; left time: 195.2609s
	iters: 200, epoch: 1 | loss: 0.2405966
	speed: 0.0632s/iter; left time: 154.2832s
Epoch: 1 cost time: 18.004290342330933
Epoch: 1, Steps: 264 | Train Loss: 0.3269685 Vali Loss: 0.2746330 Test Loss: 0.3945290
Validation loss decreased (inf --> 0.274633).  Saving model ...
Updating learning rate to 0.0001
	iters: 100, epoch: 2 | loss: 0.3052072
	speed: 0.1449s/iter; left time: 329.8502s
	iters: 200, epoch: 2 | loss: 0.1317495
	speed: 0.0626s/iter; left time: 136.2892s
Epoch: 2 cost time: 16.520211696624756
Epoch: 2, Steps: 264 | Train Loss: 0.2197954 Vali Loss: 0.2868253 Test Loss: 0.4253391
EarlyStopping counter: 1 out of 3
Updating learning rate to 5e-05
	iters: 100, epoch: 3 | loss: 0.1411438
	speed: 0.1452s/iter; left time: 292.2331s
	iters: 200, epoch: 3 | loss: 0.1338760
	speed: 0.0631s/iter; left time: 120.7719s
Epoch: 3 cost time: 16.639776468276978
Epoch: 3, Steps: 264 | Train Loss: 0.1703486 Vali Loss: 0.2900613 Test Loss: 0.4549606
EarlyStopping counter: 2 out of 3
Updating learning rate to 2.5e-05
	iters: 100, epoch: 4 | loss: 0.1420920
	speed: 0.1458s/iter; left time: 255.0376s
	iters: 200, epoch: 4 | loss: 0.1750177
	speed: 0.0632s/iter; left time: 104.2967s
Epoch: 4 cost time: 16.763848543167114
Epoch: 4, Steps: 264 | Train Loss: 0.1526386 Vali Loss: 0.2885140 Test Loss: 0.4750347
EarlyStopping counter: 3 out of 3
Early stopping
>>>>>>>testing : ETTh2_96_96_ns_Transformer_ETTh2_ftM_sl96_ll48_pl96_dm512_nh8_el2_dl1_df2048_fc1_ebtimeF_dtTrue_'Exp_h256_l2'_0<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
test 2785
test shape: (87, 32, 96, 7) (87, 32, 96, 7)
test shape: (2784, 96, 7) (2784, 96, 7)
mse:0.3945288360118866, mae:0.41980457305908203
Use GPU: cuda:0
>>>>>>>start training : ETTh2_96_96_ns_Transformer_ETTh2_ftM_sl96_ll48_pl96_dm512_nh8_el2_dl1_df2048_fc1_ebtimeF_dtTrue_'Exp_h256_l2'_1>>>>>>>>>>>>>>>>>>>>>>>>>>
train 8449
val 2785
test 2785
	iters: 100, epoch: 1 | loss: 0.2209248
	speed: 0.0653s/iter; left time: 165.9310s
	iters: 200, epoch: 1 | loss: 0.2358510
	speed: 0.0638s/iter; left time: 155.7341s
Epoch: 1 cost time: 17.02840828895569
Epoch: 1, Steps: 264 | Train Loss: 0.3301022 Vali Loss: 0.2849490 Test Loss: 0.3910953
Validation loss decreased (inf --> 0.284949).  Saving model ...
Updating learning rate to 0.0001
	iters: 100, epoch: 2 | loss: 0.2643779
	speed: 0.1497s/iter; left time: 340.8089s
	iters: 200, epoch: 2 | loss: 0.2535767
	speed: 0.0641s/iter; left time: 139.5449s
Epoch: 2 cost time: 16.864766597747803
Epoch: 2, Steps: 264 | Train Loss: 0.2152764 Vali Loss: 0.2894374 Test Loss: 0.4258763
EarlyStopping counter: 1 out of 3
Updating learning rate to 5e-05
	iters: 100, epoch: 3 | loss: 0.1445326
	speed: 0.1484s/iter; left time: 298.6893s
	iters: 200, epoch: 3 | loss: 0.1796765
	speed: 0.0643s/iter; left time: 123.0670s
Epoch: 3 cost time: 17.12177586555481
Epoch: 3, Steps: 264 | Train Loss: 0.1675327 Vali Loss: 0.2817460 Test Loss: 0.4827897
Validation loss decreased (0.284949 --> 0.281746).  Saving model ...
Updating learning rate to 2.5e-05
	iters: 100, epoch: 4 | loss: 0.1448471
	speed: 0.1531s/iter; left time: 267.8125s
	iters: 200, epoch: 4 | loss: 0.1152792
	speed: 0.0649s/iter; left time: 107.0724s
Epoch: 4 cost time: 17.29974603652954
Epoch: 4, Steps: 264 | Train Loss: 0.1513977 Vali Loss: 0.2860749 Test Loss: 0.4547649
EarlyStopping counter: 1 out of 3
Updating learning rate to 1.25e-05
	iters: 100, epoch: 5 | loss: 0.1340131
	speed: 0.1480s/iter; left time: 219.7723s
	iters: 200, epoch: 5 | loss: 0.1168449
	speed: 0.0645s/iter; left time: 89.3987s
Epoch: 5 cost time: 17.2496280670166
Epoch: 5, Steps: 264 | Train Loss: 0.1434257 Vali Loss: 0.2873148 Test Loss: 0.4631858
EarlyStopping counter: 2 out of 3
Updating learning rate to 6.25e-06
	iters: 100, epoch: 6 | loss: 0.1203539
	speed: 0.1544s/iter; left time: 188.5536s
	iters: 200, epoch: 6 | loss: 0.1027171
	speed: 0.0647s/iter; left time: 72.5008s
Epoch: 6 cost time: 17.324661016464233
Epoch: 6, Steps: 264 | Train Loss: 0.1395933 Vali Loss: 0.2859809 Test Loss: 0.4624731
EarlyStopping counter: 3 out of 3
Early stopping
>>>>>>>testing : ETTh2_96_96_ns_Transformer_ETTh2_ftM_sl96_ll48_pl96_dm512_nh8_el2_dl1_df2048_fc1_ebtimeF_dtTrue_'Exp_h256_l2'_1<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
test 2785
test shape: (87, 32, 96, 7) (87, 32, 96, 7)
test shape: (2784, 96, 7) (2784, 96, 7)
mse:0.4827895164489746, mae:0.46259817481040955
Use GPU: cuda:0
>>>>>>>start training : ETTh2_96_96_ns_Transformer_ETTh2_ftM_sl96_ll48_pl96_dm512_nh8_el2_dl1_df2048_fc1_ebtimeF_dtTrue_'Exp_h256_l2'_2>>>>>>>>>>>>>>>>>>>>>>>>>>
train 8449
val 2785
test 2785
	iters: 100, epoch: 1 | loss: 0.4084912
	speed: 0.0673s/iter; left time: 170.8999s
	iters: 200, epoch: 1 | loss: 0.2444842
	speed: 0.0665s/iter; left time: 162.3365s
Epoch: 1 cost time: 17.56477952003479
Epoch: 1, Steps: 264 | Train Loss: 0.3333128 Vali Loss: 0.2911936 Test Loss: 0.4261954
Validation loss decreased (inf --> 0.291194).  Saving model ...
Updating learning rate to 0.0001
	iters: 100, epoch: 2 | loss: 0.3382215
	speed: 0.1517s/iter; left time: 345.3431s
	iters: 200, epoch: 2 | loss: 0.2358498
	speed: 0.0658s/iter; left time: 143.2543s
Epoch: 2 cost time: 17.257097005844116
Epoch: 2, Steps: 264 | Train Loss: 0.2223898 Vali Loss: 0.3071573 Test Loss: 0.4283208
EarlyStopping counter: 1 out of 3
Updating learning rate to 5e-05
	iters: 100, epoch: 3 | loss: 0.1242966
	speed: 0.1516s/iter; left time: 305.2351s
	iters: 200, epoch: 3 | loss: 0.1247057
	speed: 0.0645s/iter; left time: 123.4376s
Epoch: 3 cost time: 17.220128059387207
Epoch: 3, Steps: 264 | Train Loss: 0.1741069 Vali Loss: 0.3256035 Test Loss: 0.4496353
EarlyStopping counter: 2 out of 3
Updating learning rate to 2.5e-05
	iters: 100, epoch: 4 | loss: 0.1237622
	speed: 0.1516s/iter; left time: 265.1746s
	iters: 200, epoch: 4 | loss: 0.1213419
	speed: 0.0649s/iter; left time: 107.0502s
Epoch: 4 cost time: 17.295646905899048
Epoch: 4, Steps: 264 | Train Loss: 0.1562201 Vali Loss: 0.3242903 Test Loss: 0.4585052
EarlyStopping counter: 3 out of 3
Early stopping
>>>>>>>testing : ETTh2_96_96_ns_Transformer_ETTh2_ftM_sl96_ll48_pl96_dm512_nh8_el2_dl1_df2048_fc1_ebtimeF_dtTrue_'Exp_h256_l2'_2<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
test 2785
test shape: (87, 32, 96, 7) (87, 32, 96, 7)
test shape: (2784, 96, 7) (2784, 96, 7)
mse:0.4261956214904785, mae:0.4425164461135864

**Comment by stillwaterman** [comment]
还有一个问题，我在上面的一次ETTh2实验中，mse:0.3945288360118866, mae:0.41980457305908203,我看您的文章中ETTh2的实验结果是mse：0.476，mae：0.458，我想问一下您是如何计算得出实验结果的？

**Comment by WenWeiTHU** [comment]
> 还有一个问题，我在上面的一次ETTh2实验中，mse:0.3945288360118866, mae:0.41980457305908203,我看您的文章中ETTh2的实验结果是mse：0.476，mae：0.458，我想问一下您是如何计算得出实验结果的？

你好，这种情况是正常的，因为ETTh2在我们的benchmark中属于一个比较小的数据集（17420个时间点），我们希望在benchmark中的各数据集上尽量统一训练超参，所以early stopping在小数据集上会起到一个防止过拟合的作用。以及，实验结果的获取：设定不同种子重复三次实验计算平均值，具体可以参考我们的附录E。

**Comment by stillwaterman** [comment]
@WenWeiTHU 非常感谢您的回复，很好地解答了我的疑问。之前做图像的，不太习惯这种训练过程，感谢您地解答，我关闭issue
